tests/browser_context_test.go

```diff
@@ -750,119 +746,58 @@ func TestBrowserContextWaitForEvent(t *testing.T) {
 			ctx, cancel := context.WithTimeout(tb.context(), 5*time.Second)
 			defer cancel()
 
-			var p1, p2 *common.Page
-			// We need to run waitForEvent in parallel to the page creation.
-			// If we run them synchronously then waitForEvent will wait
-			// indefinitely and eventually the test will timeout.
-			err = tb.run(
-				ctx,
-				func() error {
-					op := optsOrPredicateToGojaValue(t, tb, tc.optsOrPredicate)
-					resp, err := bc.WaitForEvent(tc.event, op)
-					if resp != nil {
-						var ok bool
-						p1, ok = resp.(*common.Page)
-						require.True(t, ok)
-					}
-					return err
-				},
-				func() error {
-					var err error
-					p2, err = bc.NewPage()
-					return err
-				},
+			var (
+				aboutToCallWait = make(chan bool)
+				waitDone        = make(chan bool)
+				errChan         = make(chan error)
+				p1ID, p2ID      string
 			)
 
+			go func() {
+				defer close(waitDone)
+				defer close(errChan)
+
+				var resp any
+				close(aboutToCallWait)
+				resp, err = bc.WaitForEvent(tc.event, tc.predicate, tc.timeout)
+				if resp != nil {
+					p, ok := resp.(*common.Page)
+					if !ok {
+						errChan <- errors.New("response from waitForEvent is not a page")
```

Comment:
One more point :) This goroutine won't return if the test goroutine finishes earlier without reading the `errChan` (maybe due to the `Context` timeout). Should we make the `errChan` with a single-item buffered chan to be on the safe side?

Reply:
Good point, what about reusing the context like so:

```go
						select {
						case errChan <- errors.New("response from waitForEvent is not a page"):
						case <-ctx.Done():
						}
```

Reply:
Makes sense, but it's safer to accept a one-buffered channel if we ever update this code later. These kinds of bugs are subtle and can easily go unnoticed. Have you tried using `tb.run` here? IDK if it makes sense, but it can make things easier.

Reply:
Ah, nice, i like `tb.run`, i was thinking errgroup might be a good fit 🙂  Resolved in 6855e30afdc593b67318ec34e3ad6578dae39a25

Reply:
Perfect 👏 🙇